File: go/heap/max-heap.go
package heap

type MaxHeap struct {
	Elements []int
}

func NewMaxHeap() *MaxHeap {
	return &MaxHeap{}
}

func (h *MaxHeap) Insert(data int) {
	h.Elements = append(h.Elements, data)
	newElementIndex := len(h.Elements) - 1
	parentIndex := getParentIndex(newElementIndex)
	for parentIndex != -1 && h.Elements[newElementIndex] > h.Elements[parentIndex] {
		h.Elements[newElementIndex], h.Elements[parentIndex] = h.Elements[parentIndex], h.Elements[newElementIndex]
		newElementIndex = parentIndex
		parentIndex = getParentIndex(newElementIndex)
	}
}

func (h *MaxHeap) Extract() (int, bool) {
	if len(h.Elements) == 0 {
		return 0, false
	}
	extractedElement := h.Elements[0]
	lastElementIndex := len(h.Elements) - 1
	h.Elements[0] = h.Elements[lastElementIndex]
	h.Elements = h.Elements[:lastElementIndex]
	parentIndex := 0
	leftChildIndex := getLeftChildIndex(parentIndex)
	rightChildIndex := getRightChildIndex(parentIndex)

	for leftChildIndex < len(h.Elements) && rightChildIndex < len(h.Elements) {
		maxChildIndex := h.max(leftChildIndex, rightChildIndex)
		if h.Elements[maxChildIndex] > h.Elements[parentIndex] {
			h.Elements[maxChildIndex], h.Elements[parentIndex] = h.Elements[parentIndex], h.Elements[maxChildIndex]
			parentIndex = maxChildIndex
			leftChildIndex = getLeftChildIndex(parentIndex)
			rightChildIndex = getRightChildIndex(parentIndex)
		}
	}

	if leftChildIndex < len(h.Elements) {
		if h.Elements[leftChildIndex] > h.Elements[parentIndex] {
			h.Elements[leftChildIndex], h.Elements[parentIndex] = h.Elements[parentIndex], h.Elements[leftChildIndex]
		}
	}
	return extractedElement, true
}

func (h *MaxHeap) max(leftChildIndex, rightChildIndex int) int {
	if h.Elements[leftChildIndex] > h.Elements[rightChildIndex] {
		return leftChildIndex
	}
	return rightChildIndex
}

func getParentIndex(childIndex int) int {
	if childIndex == 0 {
		return -1
	}
	return (childIndex - 1) / 2
}

func getLeftChildIndex(parentIndex int) int {
	return parentIndex*2 + 1
}

func getRightChildIndex(parentIndex int) int {
	return parentIndex*2 + 2
}

File: js/max-heap.js
class MaxHeap {
    constructor () {
        this.storage = [];
    }

    sort () {
        const result = [];

        while (this.storage.length > 0) {
            result.push(this.remove());
        }

        return result;
    }

    insert (data) {
        this.storage.push(data);
        this.heapifyUp();
    }

    heapifyUp () {
        let currentIndex = this.storage.length - 1;
        while (this.hasParent(currentIndex) === true) {
            const parentIndex = this.getParentIndex(currentIndex)
            if (this.getParent(currentIndex) < this.storage[currentIndex]) {
                this.swap(currentIndex, parentIndex);
            }
            currentIndex = parentIndex;
        }
    }

    remove () {
        if (this.storage.length === 0) return null;
        const maximum = this.storage[0];
        if (this.storage.length === 1) 
            this.storage.pop();
        else {
            this.storage[0] = this.storage.pop();
            this.heapifyDown();
        }
            
        return maximum;
    }

    heapifyDown() {
        let currentIndex = 0;
        while (this.hasLeftChild(currentIndex) === true) {
            let largestChildIndex = this.getLeftChildIndex(currentIndex);
            if (this.hasRightChild(currentIndex) && this.getRightChild(currentIndex) > this.storage[largestChildIndex])
                largestChildIndex = this.getRightChildIndex(currentIndex);
            if (this.storage[largestChildIndex] > this.storage[currentIndex]) {
                this.swap(largestChildIndex, currentIndex);
                currentIndex = largestChildIndex;
            } else {
                break;
            }
        }
    }

    swap (i, j) {
        const temp = this.storage[i];
        this.storage[i] = this.storage[j];
        this.storage[j] = temp;
    }

    getParent (childIndex) {
        return this.storage[this.getParentIndex(childIndex)];
    }

    getLeftChild (parentIndex) {
        return this.storage[this.getLeftChildIndex(parentIndex)];
    }

    getRightChild (parentIndex) {
        return this.storage[this.getRightChildIndex(parentIndex)];
    }

    hasParent (index) {
        return this.getParentIndex(index) >= 0;
    }

    hasLeftChild (index) {
        return this.getLeftChildIndex(index) < this.storage.length;
    }

    hasRightChild (index) {
        return this.getRightChildIndex(index) < this.storage.length;
    }
    
    getParentIndex (childIndex) {
        return Math.floor((childIndex - 1) / 2);
    }

    getLeftChildIndex (parentIndex) {
        return 2 * parentIndex + 1;
    }

    getRightChildIndex (parentIndex) {
        return 2 * parentIndex + 2;
    }
}
/*
                
*/

const myMaxHeap = new MaxHeap();
const arr = [10, 5, 1000, 50, 2, 8]
for (let element of arr) {
    myMaxHeap.insert(element);
}

console.log(myMaxHeap.sort());